PHEP 9999: PyHC standardization for Python time objects (#32)
heliophysicsPy/standards: pheps/phep-9999.md

```
PHEP: 9999
Title: PyHC standardization for Python time objects
Author: Nabil Freij <nabil.freij@gmail.com> <https://orcid.org/0000-0002-6253-082X>

> Discussions-To: https://github.com/heliophysicsPy/standards/pull/X

Author: ```suggestion
Discussions-To: https://github.com/heliophysicsPy/standards/pull/32
```

This meant that a some PyHC packages have had to create their own implementations of specific components like time or coordinates due to necessity.

Now that the upstream ecosystem has grown at a rapid pace, the need for PyHC package specific implementations has reduced considerably.
There are at least three areas ("time", "coordinates" and "units") that require standardization, however the focus for this PHEP is to start at "time", as that is considered the *simplest* case for standardization.

To try and address the need to reduce duplication and improve interoperability between PyHC packages, decisions have to be made about reducing the number of ways the community handles the same type of operation but it has to be considered within the wider Scientific Python ecosystem.

For example, the reason that Astropy's time framework was created was to add support for astronomical formats (e.g., Julian Date (JD), Modified JD (MJD) and precise timing (e.g., a nanosecond over a Hubble Time).
None of which is possible when using the [datetime](https://docs.python.org/3/library/datetime.html) module.

> 

Reviewer: ```suggestion

# Use case

Several independent researchers are attempting to match coronal features during an eruption observed by the Atmospheric Imaging Assembly (AIA) on the Solar Dynamics Observatory (SDO) to in situ measurements observed by Parker Solar Probe (PSP). Among these researchers are summer student interns who are using PyHC packages for the first time. This study requires careful comparisons of the times of the remote and in situ observations. In particular, the researchers must make numerous time comparisons, while taking into account transit time and spatial differences. 

The researchers decide to use SunPy (which makes use of `astropy.time.Time`) to read in and analyze remote observations from SDO/AIA. They then decide to either use SpacePy (and its `TickTock` class) or PySPEDAS (which uses `datetime64`?) to read in and analyze in situ observations from PSP. However, operations like `Time(...) - TickTock(...)` result in an exception(?).

Currently, no functionality exists(?) among PyHC packages to convert between or directly compare `TickTock`, `Time`, and `datetime64` objects. This lack of interoperability has the following consequences:

 - Students and researchers would need to learn how to use and interact with multiple different time objects, which increases the onboarding time for student interns.
 - Each researcher would need to write or acquire functionality that would convert the different time classes into a common class.(?)
 - The effort required to perform this analysis would be increased.
 - User experience is degraded.

If the ecosystem were to standardize on a common time framework (or at least one with a common API), the consequences would be:

 - Cross-disciplinary researchers would only need to learn a single API for dealing with time.
 - Time delta operations would be greatly simplified: subtracting a time object from another would produce a consistent time delta object.
 - The effort needed to perform this analysis would decrease.
 - The likelihood of unnecessary software duplication is reduced.
 - User experience when working with multiple PyHC packages would be improved.
 - The onboarding time for students would be reduced, and the suitability of this topic as a research project for student interns would be increased.
 - `astropy.time.TimeDelta` objects can be converted to `astropy.units.Quantity` objects using the `.to` method, which would enable multiplication with a velocity to get a distance.

```
Would it make sense to spell out some [use cases](https://en.wikipedia.org/wiki/Use_case) to show more about why we're hoping to standardize around time?  I attempted to draft one here, with question marks about the parts I'm uncertain about.

Reviewer 2: Hi @namurphy - I suggest trying the experiment. 

It seems that we have a proposed standard and are looking for hypothetical problems to justify it. All the packages understand `datetime`, and the hypothetical does not indicate exactly where `datetime` is insufficient. I expect one could come up with a concrete example that the standard would have fixed, but it will likely be an uncommon use case - if so, we should debate if the burden of the standard on everyone is worth simplifying an uncommon use case. 

I'd really like to see the participants try to use other packages together and study the package's source code. Five ideas for standards or cross-cutting projects will probably result, and I doubt this one involving `AstroPy.Time` will be one of them. I am very reluctant to propose a standard based on hypothetical or contrived problems with claims that may prove wrong if one tries to solve the problem. There should be a PHEP for time, but there should not be speculation to justify it. Also, the implications should be researched (again, without speculations).

**Much more research is needed. I don't see the rush to have this voted on.**

I've started with an ISO 8601 timestamp of interest and used SpacePy, SunPy, PySPEDAS, and SpiceyPy to do a coordinate transform. There may have been a time standard that would have simplified things for me - if I could have bypassed the custom time objects and just passed an ISO 8601 string to the transform functions, things would have been much easier. In my problem, there was no need to expose the package's time abstraction to the user. Several packages require the user to understand custom time and/or coordinate objects (or don't allow dimensionless vectors) to do something conceptually simple - pass a timestamp and a vector and get a new vector. I don't know if there is an actual proposal here - what is there to say beyond "make simple things easy to do" or "don't require the scientist to understand a software abstraction if unnecessary?"

As a result of this experience, I submitted an issue requesting a keyword in the `hapiclient` function to request the output as `datetimes` instead of the default of the raw strings served by the HAPI server. There is no reason for the general user to learn and use a separate function, `hapitime2datetime`, to convert from the raw ISO 8601 strings to `datetimes`. In the next major release, we'll make this the default behavior because this is what most want.

Author: Nabil Freij <nabil.freij@gmail.com> <https://orcid.org/0000-0002-6253-082X>
Discussions-To: https://github.com/heliophysicsPy/standards/pull/X
Revision: 1
Status: Draft
Type: Standards Track
Content-Type: text/markdown; charset=UTF-8; variant=CommonMark
Created: 08-July-2024
Post-History: 08-July-2024
```

# Abstract

This PHEP recommends that all projects across the PyHC ecosystem adopt using `astropy.time.Time` as the foundational class for all time objects.
This process would involve rewriting existing PyHC core packages replace any use of datetime or custom time objects with `astropy.time.Time`.

As this is a major change, the transition needs to be carefully managed and will be broken into stages:

1. Existing core packages will be audited to see how complex their time objects are.
2. Plans to be drawn up on how to tackle all code changes.
3. Funding to be found for this work to go ahead.

For any new packages that are being developed or created, they are recommended to use `astropy.time.Time` from the beginning.

# Motivation

One of the strategic goals of the PyHC Project is to "Coordinate development across projects to minimize duplication of effort".

> However, all of the core PyHC packages were created not only before the existence of PyHC.
> But also before many core libraries existed within the wider Scientiﬁc Python ecosystem (from `numpy` to `scipy` to `astropy` to `spiceypy` etc) meant that all PyHC packages had to create their own implementations of time (and other components like coordinates) due to necessity.

Reviewer: ```suggestion
However, all of the core PyHC packages were created not only before the existence of PyHC,
but also before many core libraries existed within the wider Scientific Python ecosystem (from `numpy` to `scipy` to `astropy` to `spiceypy` etc). This meant that all PyHC packages had to create their own implementations of time (and other components like coordinates) due to necessity.
```

All that does is slow them down.

One of the most controversial events within the Linux ecosystem (in the authors short lifetime) was the adoption of [systemd](https://systemd.io/).
Many of the arguments around that time where similar, it reduces innovation, its too complex, users have to learn a new system.
However, the maintainers of each major linux distribution embraced systemd as it provided a solid foundation to reduce their own maintenance burden but also having one system meant that one could find detailed documentation or common problems and how to solve them via an internet search.

The adoption of `astropy` is in the authors opinion, the same.
It provides a new foundation that not only reduces the amount of code within PyHC that handles the same functionality, it allows users to focus on the parts of their work that matter.
Not only would adding more features to `astropy` benefit the astronomy community, it would in turn benefit the PyHC ecosystem.

# Specification

A dedicated implementation is most likely out of scope for a PHEP, as each package will require bespoke changes after an audit.

> However, there will be a few common strategies that can be summarized below:

Reviewer: Adding the strategies for making this effort is really helpful, so thank you for adding it!

- For libraries that provide datetime as inputs, they need to add explicit `astropy.time.Time` support as inputs like for example searching for data.
- For libraries that provide a custom time object, the deprecated of such an object and support for  `astropy.time.Time` would on the surface be simplest.
  Another choice is to rebase the existing time object onto `astropy.time.Time` to avoid breaking existing user code directly.

It is important to note that for new time formats, they do not need to added to `astropy`, they can be in the same PyHC package as they are currently.
The time framework has been designed to be extendable without adding code changes to `astropy`.

As the merger of this PHEP implies that the PyHC ecosystem are *planning* on standardizing on `astropy.time.Time`, there will be need to be a ecosystem wide plan to enable this transition.

One such transition plan is as follows:

- Review of the existing core PyHC packages to figure out:
  - What time objects are used?
  - How would adding `astropy.time.Time` support as function inputs work?
  - Figure out missing time formats
  - Figure out missing features from `astropy.time.Time`

> - Figure out how to deal with `astropy.time.Time` lack of support in pandas

Reviewer: ```suggestion
- Figure out how to deal with `astropy.time.Time` lack of support in pandas and polars
```
In case anyone hasn't heard of it, [polars](https://pola.rs/) is similar to pandas, except that it is written in Rust for better performance.

However there might be more of a concern for more tightly controlled operational environments which would need to be investigated.

# How to Teach This

- Video meetings to help developers of core packages learn `astropy.time` from either people familiar with `astropy.time` or the astropy developers themselves.
- [`astropy.time` documentation is pretty extensive.](https://docs.astropy.org/en/stable/time/)
- Creation of migration guides to help users update their existing code for each library.

# Reference Implementation

Using `astropy.time` won't involve having one reference implementation.
However, both `sunpy` and `plasmapy` use `astropy.time.Time` internally and `sunpy` contains examples of extending `astropy.time.Time` to support extra time formats but also datetime string parsing.

- [New time epoch](https://github.com/sunpy/sunpy/blob/main/sunpy/time/timeformats.py)
- [New datetime string formats](https://github.com/sunpy/sunpy/blob/main/sunpy/time/time.py#L37)

# Rejected Ideas

None so far.

# Open Issues

1. What should the time frame be for package transitions?

> 2. How do we handle `astropy.time` lack of support in pandas?

Reviewer: ```suggestion
2. How do we handle `astropy.time` lack of support in pandas and polars?
```
